scripts/generate-conda-envs.py

```diff
@@ -0,0 +1,357 @@
+    def sections(self) -> Tuple[SectionConfig, ...]:
+        return (
+            self.cuda,
+            self.build,
+            self.runtime,
+            self.tests,
+            self.docs,
+        )
+
+    @property
+    def cuda(self) -> CUDAConfig:
+        return CUDAConfig(self.ctk)
+
+    @property
+    def build(self) -> BuildConfig:
+        return BuildConfig(self.compilers, self.openmpi)
+
+    @property
+    def runtime(self) -> RuntimeConfig:
+        return RuntimeConfig()
+
+    @property
+    def tests(self) -> TestsConfig:
+        return TestsConfig()
+
+    @property
+    def docs(self) -> DocsConfig:
+        return DocsConfig()
+
+    @property
+    def filename(self) -> str:
+        python = f"py{self.python.replace('.', '')}"
+        return f"environment-{self.use}-{self.os}-{python}{self.cuda}{self.build}.yaml"  # noqa
+
+
+# --- Setup -------------------------------------------------------------------
+
+PYTHON_VERSIONS = ("3.8", "3.9", "3.10")
+
+CTK_VERSIONS = (
+    "none",
+    "10.2",
+    "11.0",
+    "11.1",
+    "11.2",
+    "11.3",
+    "11.4",
+    "11.5",
+    "11.6",
+    "11.7",
+)
+
+OS_NAMES: Tuple[OSType, ...] = ("linux", "osx")
+
+
+ENV_TEMPLATE = Template(
+    """
+name: legate-{{ use }}
+channels:
+  - conda-forge
+dependencies:
+
+  - python={{ python }}
+  {% if conda_sections %}
+  {% for section in conda_sections %}
+
+  # {{ section.header }}
+  {% for req in section.conda %}
+  - {{ req }}
+  {% endfor %}
+  {% endfor %}
+  {% endif %}
+  {% if pip_sections %}
+
+  - pip
+  - pip:
+      {% for section in pip_sections %}
+
+      # {{ section.header }}
+      {% for req in section.pip %}
+      - {{ req }}
+      {% endfor %}
+      {% endfor %}
+  {% endif %}
+""",
+    trim_blocks=True,
+    lstrip_blocks=True,
+)
+
+ALL_CONFIGS = [
+    EnvConfig("test", python, "linux", ctk, compilers, openmpi)
+    for python in PYTHON_VERSIONS
+    for ctk in CTK_VERSIONS
+    for compilers in (True, False)
+    for openmpi in (True, False)
+] + [
+    EnvConfig("test", python, "darwin", "none", compilers, openmpi)
+    for python in PYTHON_VERSIONS
+    for compilers in (True, False)
+    for openmpi in (True, False)
+]
+
+# --- Code --------------------------------------------------------------------
+
+if __name__ == "__main__":
+
+    import sys
+    from argparse import ArgumentParser
+
+    parser = ArgumentParser()
+    parser.add_argument(
+        "--python",
+        choices=PYTHON_VERSIONS,
+        default=None,
+        help="Python version to generate for, (default: all python versions)",
+    )
+    parser.add_argument(
+        "--ctk",
+        choices=CTK_VERSIONS,
+        default=None,
+        dest="ctk_version",
+        help="CTK version to generate for (default: all CTK versions)",
+    )
+    parser.add_argument(
+        "--os",
+        choices=OS_NAMES,
+        default=None,
+        help="OS to generate for (default: all OSes)",
+    )
+    parser.add_argument(
```

Comment:
`BooleanOptionalAction` is not available in python 3.8, so I did it like so. This works, but the help message doesn't look that great:

```
...
  --os {linux,osx}      OS to generate for (default: all OSes)
  --compilers           Whether to include conda compilers or not (default: both)
  --no-compilers        Whether to include conda compilers or not (default: both)
  --openmpi             Whether to include openmpi or not (default: both)
  --no-openmpi          Whether to include openmpi or not (default: both)
  ```

Reply:
If `BooleanOptionalAction` is small/self-contained we could just lift it wholesale

Reply:
We actually have a `BooleanFlag` already in `install.py`.